### DISCUSSION TITLE
No module named 'eth_typing' [solved by instaling Microsoft Visual C++ 14.0 or greater]

### DISCUSSION THREAD

**Discussion by GroverTa1914** [coccoinomane/crabada.py#119] [Q&A] [answered]
NeverMind.
I was missing the C++ library need for the pip install.

Thanks,
Tim

Hello,
I was able to get it working on one of my older development laptops.
But when I try a new laptop windows 10 i am getting this error.

Traceback (most recent call last):
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.1264.0_x64__qbz5n2kfra8p0\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.1264.0_x64__qbz5n2kfra8p0\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Users\Parad\Documents\crabada.py-master\bin\mining\run.py", line 11, in <module>
    from eth_typing import Address
ModuleNotFoundError: No module named 'eth_typing'

am I doing something wrong.
Please help.
Thank you for everything.

Tim

Here is the install:

C:\Users\Parad\Documents\crabada.py-master>pip install -r requirements.txt
Collecting python-dotenv==0.19.*
  Downloading python_dotenv-0.19.2-py2.py3-none-any.whl (17 kB)
Collecting requests==2.26.*
  Downloading requests-2.26.0-py2.py3-none-any.whl (62 kB)
     ---------------------------------------- 62.3/62.3 KB 3.3 MB/s eta 0:00:00
Collecting types-requests==2.26.*
  Downloading types_requests-2.26.3-py3-none-any.whl (22 kB)
Collecting web3==5.25.*
  Downloading web3-5.25.0-py3-none-any.whl (485 kB)
     ---------------------------------------- 485.4/485.4 KB 15.3 MB/s eta 0:00:00
Collecting eth-typing==2.2.*
  Downloading eth_typing-2.2.2-py3-none-any.whl (6.2 kB)
Collecting urllib3<1.27,>=1.21.1
  Downloading urllib3-1.26.9-py2.py3-none-any.whl (138 kB)
     ---------------------------------------- 139.0/139.0 KB ? eta 0:00:00
Collecting idna<4,>=2.5
  Downloading idna-3.3-py3-none-any.whl (61 kB)
     ---------------------------------------- 61.2/61.2 KB ? eta 0:00:00
Collecting certifi>=2017.4.17
  Downloading certifi-2022.5.18.1-py3-none-any.whl (155 kB)
     ---------------------------------------- 155.2/155.2 KB 9.7 MB/s eta 0:00:00
Collecting charset-normalizer~=2.0.0
  Downloading charset_normalizer-2.0.12-py3-none-any.whl (39 kB)
Collecting hexbytes<1.0.0,>=0.1.0
  Downloading hexbytes-0.2.2-py3-none-any.whl (6.1 kB)
Collecting ipfshttpclient==0.8.0a2
  Downloading ipfshttpclient-0.8.0a2-py3-none-any.whl (82 kB)
     ---------------------------------------- 82.6/82.6 KB ? eta 0:00:00
Collecting eth-account<0.6.0,>=0.5.6
  Downloading eth_account-0.5.7-py3-none-any.whl (101 kB)
     ---------------------------------------- 101.8/101.8 KB ? eta 0:00:00
Collecting websockets<10,>=9.1
  Downloading websockets-9.1.tar.gz (76 kB)
     ---------------------------------------- 76.7/76.7 KB ? eta 0:00:00
  Preparing metadata (setup.py) ... done
Collecting eth-hash[pycryptodome]<1.0.0,>=0.2.0
  Downloading eth_hash-0.3.2-py3-none-any.whl (8.8 kB)
Collecting pywin32>=223
  Downloading pywin32-304-cp310-cp310-win_amd64.whl (12.1 MB)
     ---------------------------------------- 12.1/12.1 MB 20.5 MB/s eta 0:00:00
Collecting jsonschema<4.0.0,>=3.2.0
  Downloading jsonschema-3.2.0-py2.py3-none-any.whl (56 kB)
     ---------------------------------------- 56.3/56.3 KB ? eta 0:00:00
Collecting eth-utils<2.0.0,>=1.9.5
  Downloading eth_utils-1.10.0-py3-none-any.whl (24 kB)
Collecting aiohttp<4,>=3.7.4.post0
  Downloading aiohttp-3.8.1-cp310-cp310-win_amd64.whl (555 kB)
     ---------------------------------------- 555.1/555.1 KB 17.6 MB/s eta 0:00:00
Collecting eth-abi<3.0.0,>=2.0.0b6
  Downloading eth_abi-2.1.1-py3-none-any.whl (27 kB)
Collecting protobuf<4,>=3.10.0
  Downloading protobuf-3.20.1-cp310-cp310-win_amd64.whl (903 kB)
     ---------------------------------------- 903.8/903.8 KB 28.8 MB/s eta 0:00:00
Collecting lru-dict<2.0.0,>=1.1.6
  Downloading lru-dict-1.1.7.tar.gz (10 kB)
  Preparing metadata (setup.py) ... done
Collecting multiaddr>=0.0.7
  Downloading multiaddr-0.0.9-py2.py3-none-any.whl (16 kB)
Collecting multidict<7.0,>=4.5
  Downloading multidict-6.0.2-cp310-cp310-win_amd64.whl (27 kB)
Collecting attrs>=17.3.0
  Downloading attrs-21.4.0-py2.py3-none-any.whl (60 kB)
     ---------------------------------------- 60.6/60.6 KB 3.1 MB/s eta 0:00:00
Collecting async-timeout<5.0,>=4.0.0a3
  Downloading async_timeout-4.0.2-py3-none-any.whl (5.8 kB)
Collecting frozenlist>=1.1.1
  Downloading frozenlist-1.3.0-cp310-cp310-win_amd64.whl (33 kB)
Collecting yarl<2.0,>=1.0
  Downloading yarl-1.7.2-cp310-cp310-win_amd64.whl (122 kB)
     ---------------------------------------- 122.2/122.2 KB 7.0 MB/s eta 0:00:00
Collecting aiosignal>=1.1.2
  Downloading aiosignal-1.2.0-py3-none-any.whl (8.2 kB)
Collecting parsimonious<0.9.0,>=0.8.0
  Downloading parsimonious-0.8.1.tar.gz (45 kB)
     ---------------------------------------- 45.1/45.1 KB ? eta 0:00:00
  Preparing metadata (setup.py) ... done
Collecting eth-keyfile<0.6.0,>=0.5.0
  Downloading eth_keyfile-0.5.1-py3-none-any.whl (8.3 kB)
Collecting eth-rlp<2,>=0.1.2
  Downloading eth_rlp-0.3.0-py3-none-any.whl (5.0 kB)
Collecting bitarray<1.3.0,>=1.2.1
  Downloading bitarray-1.2.2.tar.gz (48 kB)
     ---------------------------------------- 48.2/48.2 KB 2.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
Collecting rlp<3,>=1.0.0
  Downloading rlp-2.0.1-py2.py3-none-any.whl (20 kB)
Collecting eth-keys<0.4.0,>=0.3.4
  Downloading eth_keys-0.3.4-py3-none-any.whl (21 kB)
Collecting pycryptodome<4,>=3.6.6
  Downloading pycryptodome-3.14.1-cp35-abi3-win_amd64.whl (1.8 MB)
     ---------------------------------------- 1.8/1.8 MB 29.0 MB/s eta 0:00:00
Collecting cytoolz<1.0.0,>=0.10.1
  Downloading cytoolz-0.11.2.tar.gz (481 kB)
     ---------------------------------------- 481.0/481.0 KB 31.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
Requirement already satisfied: setuptools in c:\program files\windowsapps\pythonsoftwarefoundation.python.3.10_3.10.1264.0_x64__qbz5n2kfra8p0\lib\site-packages (from jsonschema<4.0.0,>=3.2.0->web3==5.25.*->-r requirements.txt (line 4)) (58.1.0)
Collecting pyrsistent>=0.14.0
  Downloading pyrsistent-0.18.1-cp310-cp310-win_amd64.whl (61 kB)
     ---------------------------------------- 61.6/61.6 KB 3.2 MB/s eta 0:00:00
Collecting six>=1.11.0
  Downloading six-1.16.0-py2.py3-none-any.whl (11 kB)
Collecting toolz>=0.8.0
  Downloading toolz-0.11.2-py3-none-any.whl (55 kB)
     ---------------------------------------- 55.8/55.8 KB ? eta 0:00:00
Collecting eth-rlp<2,>=0.1.2
  Downloading eth_rlp-0.2.1-py3-none-any.whl (5.0 kB)
Collecting varint
  Downloading varint-1.0.2.tar.gz (1.9 kB)
  Preparing metadata (setup.py) ... done
Collecting netaddr
  Downloading netaddr-0.8.0-py2.py3-none-any.whl (1.9 MB)
     ---------------------------------------- 1.9/1.9 MB 24.2 MB/s eta 0:00:00
Collecting base58
  Downloading base58-2.1.1-py3-none-any.whl (5.6 kB)
Using legacy 'setup.py install' for lru-dict, since package 'wheel' is not installed.
Using legacy 'setup.py install' for websockets, since package 'wheel' is not installed.
Using legacy 'setup.py install' for bitarray, since package 'wheel' is not installed.
Using legacy 'setup.py install' for cytoolz, since package 'wheel' is not installed.
Using legacy 'setup.py install' for parsimonious, since package 'wheel' is not installed.
Using legacy 'setup.py install' for varint, since package 'wheel' is not installed.
Installing collected packages: varint, types-requests, pywin32, netaddr, lru-dict, bitarray, websockets, urllib3, toolz, six, python-dotenv, pyrsistent, pycryptodome, protobuf, multidict, idna, hexbytes, frozenlist, eth-typing, eth-hash, charset-normalizer, certifi, base58, attrs, async-timeout, yarl, requests, parsimonious, multiaddr, jsonschema, cytoolz, aiosignal, ipfshttpclient, eth-utils, aiohttp, rlp, eth-keys, eth-abi, eth-rlp, eth-keyfile, eth-account, web3
  Running setup.py install for varint ... done
  WARNING: The script netaddr.exe is installed in 'C:\Users\Parad\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\Scripts' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Running setup.py install for lru-dict ... error
  error: subprocess-exited-with-error

  × Running setup.py install for lru-dict did not run successfully.
  │ exit code: 1
  ╰─> [5 lines of output]
      running install
      running build
      running build_ext
      building 'lru' extension
      error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: legacy-install-failure

× Encountered error while trying to install package.
╰─> lru-dict

note: This is an issue with the package mentioned above, not pip.
hint: See above for output from the failure.
WARNING: You are using pip version 22.0.4; however, version 22.1 is available.
You should consider upgrading via the 'C:\Users\Parad\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip' command.

C:\Users\Parad\Documents\crabada.py-master>python -m bin.mining.run 0x73A26C4bA35d1eAecb119b3b26EeAc9F67352e9e
Traceback (most recent call last):
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.1264.0_x64__qbz5n2kfra8p0\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.1264.0_x64__qbz5n2kfra8p0\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Users\Parad\Documents\crabada.py-master\bin\mining\run.py", line 11, in <module>
    from eth_typing import Address
ModuleNotFoundError: No module named 'eth_typing'

C:\Users\Parad\Documents\crabada.py-master>python -m bin.mining.run 0x73A26C4bA35d1eAecb119b3b26EeAc9F67352e9e
Traceback (most recent call last):
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.1264.0_x64__qbz5n2kfra8p0\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.1264.0_x64__qbz5n2kfra8p0\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Users\Parad\Documents\crabada.py-master\bin\mining\run.py", line 11, in <module>
    from eth_typing import Address
ModuleNotFoundError: No module named 'eth_typing'

C:\Users\Parad\Documents\crabada.py-master>python -m bin.mining.run 73A26C4bA35d1eAecb119b3b26EeAc9F67352e9e
Traceback (most recent call last):
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.1264.0_x64__qbz5n2kfra8p0\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.1264.0_x64__qbz5n2kfra8p0\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Users\Parad\Documents\crabada.py-master\bin\mining\run.py", line 11, in <module>
    from eth_typing import Address
ModuleNotFoundError: No module named 'eth_typing'

C:\Users\Parad\Documents\crabada.py-master>pip install -r requirements.txt
Collecting python-dotenv==0.19.*
  Using cached python_dotenv-0.19.2-py2.py3-none-any.whl (17 kB)
Collecting requests==2.26.*
  Using cached requests-2.26.0-py2.py3-none-any.whl (62 kB)
Requirement already satisfied: types-requests==2.26.* in c:\users\parad\appdata\local\packages\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\localcache\local-packages\python310\site-packages (from -r requirements.txt (line 3)) (2.26.3)
Collecting web3==5.25.*
  Using cached web3-5.25.0-py3-none-any.whl (485 kB)
Collecting eth-typing==2.2.*
  Using cached eth_typing-2.2.2-py3-none-any.whl (6.2 kB)
Collecting certifi>=2017.4.17
  Using cached certifi-2022.5.18.1-py3-none-any.whl (155 kB)
Collecting idna<4,>=2.5
  Using cached idna-3.3-py3-none-any.whl (61 kB)
Collecting urllib3<1.27,>=1.21.1
  Using cached urllib3-1.26.9-py2.py3-none-any.whl (138 kB)
Collecting charset-normalizer~=2.0.0
  Using cached charset_normalizer-2.0.12-py3-none-any.whl (39 kB)
Requirement already satisfied: pywin32>=223 in c:\users\parad\appdata\local\packages\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\localcache\local-packages\python310\site-packages (from web3==5.25.*->-r requirements.txt (line 4)) (304)
Collecting eth-account<0.6.0,>=0.5.6
  Using cached eth_account-0.5.7-py3-none-any.whl (101 kB)
Collecting hexbytes<1.0.0,>=0.1.0
  Using cached hexbytes-0.2.2-py3-none-any.whl (6.1 kB)
Collecting eth-utils<2.0.0,>=1.9.5
  Using cached eth_utils-1.10.0-py3-none-any.whl (24 kB)
Collecting eth-abi<3.0.0,>=2.0.0b6
  Using cached eth_abi-2.1.1-py3-none-any.whl (27 kB)
Collecting jsonschema<4.0.0,>=3.2.0
  Using cached jsonschema-3.2.0-py2.py3-none-any.whl (56 kB)
Collecting ipfshttpclient==0.8.0a2
  Using cached ipfshttpclient-0.8.0a2-py3-none-any.whl (82 kB)
Collecting lru-dict<2.0.0,>=1.1.6
  Using cached lru-dict-1.1.7.tar.gz (10 kB)
  Preparing metadata (setup.py) ... done
Collecting protobuf<4,>=3.10.0
  Using cached protobuf-3.20.1-cp310-cp310-win_amd64.whl (903 kB)
Collecting websockets<10,>=9.1
  Using cached websockets-9.1.tar.gz (76 kB)
  Preparing metadata (setup.py) ... done
Collecting aiohttp<4,>=3.7.4.post0
  Using cached aiohttp-3.8.1-cp310-cp310-win_amd64.whl (555 kB)
Collecting eth-hash[pycryptodome]<1.0.0,>=0.2.0
  Using cached eth_hash-0.3.2-py3-none-any.whl (8.8 kB)
Collecting multiaddr>=0.0.7
  Using cached multiaddr-0.0.9-py2.py3-none-any.whl (16 kB)
Collecting multidict<7.0,>=4.5
  Using cached multidict-6.0.2-cp310-cp310-win_amd64.whl (27 kB)
Collecting attrs>=17.3.0
  Using cached attrs-21.4.0-py2.py3-none-any.whl (60 kB)
Collecting yarl<2.0,>=1.0
  Using cached yarl-1.7.2-cp310-cp310-win_amd64.whl (122 kB)
Collecting async-timeout<5.0,>=4.0.0a3
  Using cached async_timeout-4.0.2-py3-none-any.whl (5.8 kB)
Collecting aiosignal>=1.1.2
  Using cached aiosignal-1.2.0-py3-none-any.whl (8.2 kB)
Collecting frozenlist>=1.1.1
  Using cached frozenlist-1.3.0-cp310-cp310-win_amd64.whl (33 kB)
Collecting parsimonious<0.9.0,>=0.8.0
  Using cached parsimonious-0.8.1.tar.gz (45 kB)
  Preparing metadata (setup.py) ... done
Collecting rlp<3,>=1.0.0
  Using cached rlp-2.0.1-py2.py3-none-any.whl (20 kB)
Collecting eth-rlp<2,>=0.1.2
  Using cached eth_rlp-0.3.0-py3-none-any.whl (5.0 kB)
Collecting eth-keys<0.4.0,>=0.3.4
  Using cached eth_keys-0.3.4-py3-none-any.whl (21 kB)
Collecting bitarray<1.3.0,>=1.2.1
  Using cached bitarray-1.2.2.tar.gz (48 kB)
  Preparing metadata (setup.py) ... done
Collecting eth-keyfile<0.6.0,>=0.5.0
  Using cached eth_keyfile-0.5.1-py3-none-any.whl (8.3 kB)
Collecting pycryptodome<4,>=3.6.6
  Using cached pycryptodome-3.14.1-cp35-abi3-win_amd64.whl (1.8 MB)
Collecting cytoolz<1.0.0,>=0.10.1
  Using cached cytoolz-0.11.2.tar.gz (481 kB)
  Preparing metadata (setup.py) ... done
Collecting six>=1.11.0
  Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
Requirement already satisfied: setuptools in c:\program files\windowsapps\pythonsoftwarefoundation.python.3.10_3.10.1264.0_x64__qbz5n2kfra8p0\lib\site-packages (from jsonschema<4.0.0,>=3.2.0->web3==5.25.*->-r requirements.txt (line 4)) (58.1.0)
Collecting pyrsistent>=0.14.0
  Using cached pyrsistent-0.18.1-cp310-cp310-win_amd64.whl (61 kB)
Collecting toolz>=0.8.0
  Using cached toolz-0.11.2-py3-none-any.whl (55 kB)
Collecting eth-rlp<2,>=0.1.2
  Using cached eth_rlp-0.2.1-py3-none-any.whl (5.0 kB)
Requirement already satisfied: varint in c:\users\parad\appdata\local\packages\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\localcache\local-packages\python310\site-packages (from multiaddr>=0.0.7->ipfshttpclient==0.8.0a2->web3==5.25.*->-r requirements.txt (line 4)) (1.0.2)
Requirement already satisfied: netaddr in c:\users\parad\appdata\local\packages\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\localcache\local-packages\python310\site-packages (from multiaddr>=0.0.7->ipfshttpclient==0.8.0a2->web3==5.25.*->-r requirements.txt (line 4)) (0.8.0)
Collecting base58
  Using cached base58-2.1.1-py3-none-any.whl (5.6 kB)
Using legacy 'setup.py install' for lru-dict, since package 'wheel' is not installed.
Using legacy 'setup.py install' for websockets, since package 'wheel' is not installed.
Using legacy 'setup.py install' for bitarray, since package 'wheel' is not installed.
Using legacy 'setup.py install' for cytoolz, since package 'wheel' is not installed.
Using legacy 'setup.py install' for parsimonious, since package 'wheel' is not installed.
Installing collected packages: lru-dict, bitarray, websockets, urllib3, toolz, six, python-dotenv, pyrsistent, pycryptodome, protobuf, multidict, idna, hexbytes, frozenlist, eth-typing, eth-hash, charset-normalizer, certifi, base58, attrs, async-timeout, yarl, requests, parsimonious, multiaddr, jsonschema, cytoolz, aiosignal, ipfshttpclient, eth-utils, aiohttp, rlp, eth-keys, eth-abi, eth-rlp, eth-keyfile, eth-account, web3
  Running setup.py install for lru-dict ... error
  error: subprocess-exited-with-error

  × Running setup.py install for lru-dict did not run successfully.
  │ exit code: 1
  ╰─> [5 lines of output]
      running install
      running build
      running build_ext
      building 'lru' extension
      error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: legacy-install-failure

× Encountered error while trying to install package.
╰─> lru-dict

note: This is an issue with the package mentioned above, not pip.
hint: See above for output from the failure.
WARNING: You are using pip version 22.0.4; however, version 22.1 is available.
You should consider upgrading via the 'C:\Users\Parad\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip' command.

**Comment by coccoinomane** [comment, marked as answer]
Thank you @GroverTa1914 for sharing your problem and your solution:

```
I was missing the C++ library need for the pip install.
```

More specifically, I think the issue was solved by installing **Microsoft Visual C++ 14.0** or greater (get it with "Microsoft C++ Build Tools").

Let us know if you have more issues  🙂

Cheers,
Cocco
